sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -278,27 +280,159 @@ class RocksDB(
   // We send snapshots that needs to be uploaded by the maintenance thread to this queue
   private val snapshotsToUploadQueue = new ConcurrentLinkedQueue[RocksDBSnapshot]()
 
+  /**
+   * Read the lineage from the changelog files. It first get the changelog reader
+   * of the correct changelog version and then read the lineage information from the file.
+   * The changelog file is named as version_stateStoreCkptId.changelog
+   * @param version version of the changelog file, used to load changelog file.
+   * @param stateStoreCkptId uniqueId of the changelog file, used to load changelog file.
+   * @return
+   */
+  private def getLineageFromChangelogFile(
+      version: Long,
+      stateStoreCkptId: Option[String]): Array[LineageItem] = {
+    var changelogReader: StateStoreChangelogReader = null
+    var currLineage: Array[LineageItem] = Array.empty
+    try {
+      changelogReader = fileManager.getChangelogReader(version, stateStoreCkptId)
+      currLineage = changelogReader.lineage
+      logInfo(log"Loading lineage: " +
+        log"${MDC(LogKeys.LINEAGE, lineageManager)} from " +
+        log"changelog version: ${MDC(LogKeys.VERSION_NUM, version)} " +
+        log"uniqueId: ${MDC(LogKeys.UUID, stateStoreCkptId.getOrElse(""))}.")
+    } finally {
+      if (changelogReader != null) {
+        changelogReader.closeIfNeeded()
+      }
+    }
+    currLineage
+  }
+
+
   /**
    * Load the given version of data in a native RocksDB instance.
    * Note that this will copy all the necessary file from DFS to local disk as needed,
    * and possibly restart the native RocksDB instance.
    */
-  def load(
+  private def loadV2(
+        version: Long,
+        stateStoreCkptId: Option[String],
+        readOnly: Boolean = false): RocksDB = {
+  // An array contains lineage information from [snapShotVersion, version] (inclusive both end)
+  var currVersionLineage: Array[LineageItem] = lineageManager.getLineage()
```

Comment:
Can it be moved inside the if below?

Reply:
No this is intentionally put here. In the case of (loadedVersion==version), we don't need to reload everything from DFS, which is the most common case in reality. Then we need to reuse the existing lineage from memory.

Reply:
I see. It makes sense.
Here is an issue though. Now the assumption is that `currVersionLineage` is for `loadedVersion`. It is because `currVersionLIneage` is set in line 339 and 345, but `loadedVersion` is reset in line 382. Exceptions thrown in the middle would make these two inconsistent. Then you need to do the reset to two in the same place. Also it is better to rename function `lineageManager.getLineage()` to `lineageManager.getLineageForCurVersion()` or something like that.

Reply:
Can you elaborate more? I think every exception thrown would invalidate data from the memory: 
https://github.com/apache/spark/blob/abe6056bf8e10e52081a0becad598f2f5ff22e06/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala#L401-L409

Renamed the function

Reply:
It's a good point. Then there is no correctness issue here, which is good. It still feel better for those final update to happen in the same place, which is less prone to issues when we change the code in the future and do some early return, or handle exceptions differently.